interfaces/apparmor/backend_test.go

```diff
@@ -0,0 +1,509 @@
+	err = s.repo.AddInterface(s.iface)
+	c.Assert(err, IsNil)
+}
+
+func (s *backendSuite) TearDownTest(c *C) {
+	for _, cmd := range s.cmds {
+		cmd.Restore()
+	}
+	dirs.SetRootDir("/")
+}
+
+// Tests for Configure() and Deconfigure()
+
+func (s *backendSuite) TestInstallingSnapWritesAndLoadsProfiles(c *C) {
+	developerMode := false
+	s.installSnap(c, developerMode, `
+name: samba
+version: version
+developer: acme
+apps:
+    smbd:
+`)
+	profile := filepath.Join(s.backend.Directory(), "snap.samba.smbd")
+	// file called "snap.sambda.smbd" was created
+	_, err := os.Stat(profile)
+	c.Check(err, IsNil)
+	// apparmor_parser was was used to load that file
+	c.Check(s.cmds["apparmor_parser"].Calls(), DeepEquals, []string{
+		fmt.Sprintf("--replace --write-cache -O no-expr-simplify --cache-loc=/var/cache/apparmor %s", profile),
+	})
+}
+
+func (s *backendSuite) TestSecurityIsStable(c *C) {
+	const yaml = `
+name: samba
+version: version
+developer: acme
+apps:
+    smbd:
+`
+	for _, developerMode := range []bool{true, false} {
+		snapInfo := s.installSnap(c, developerMode, yaml)
+		s.cmds["apparmor_parser"].ForgetCalls()
+		err := s.backend.Configure(snapInfo, s.repo, developerMode)
+		c.Assert(err, IsNil)
+		// profiles are not re-compiled or re-loaded when nothing changes
+		c.Check(s.cmds["apparmor_parser"].Calls(), HasLen, 0)
+	}
+}
+
+func (s *backendSuite) TestRemovingSnapRemovesAndUnloadsProfiles(c *C) {
+	const yaml = `
+name: samba
+version: 1
+developer: acme
+apps:
+    smbd:
+`
+	for _, developerMode := range []bool{true, false} {
+		snapInfo := s.installSnap(c, developerMode, yaml)
+		s.cmds["apparmor_parser"].ForgetCalls()
+		s.removeSnap(c, snapInfo)
+		profile := filepath.Join(s.backend.Directory(), "snap.samba.smbd")
+		// file called "snap.sambda.smbd" was removed
+		_, err := os.Stat(profile)
+		c.Check(os.IsNotExist(err), Equals, true)
+		// apparmor_parser was was used to unload the profile
+		c.Check(s.cmds["apparmor_parser"].Calls(), DeepEquals, []string{
+			"--remove snap.samba.smbd",
+		})
+	}
+}
+
+func (s *backendSuite) TestUpdatingSnapMakesNeccesaryChanges(c *C) {
+	const before = `
+name: samba
+version: 1
+developer: acme
+apps:
+    smbd:
+`
+	const after = `
+name: samba
+version: 2
+developer: acme
+apps:
+    smbd:
+`
+	for _, developerMode := range []bool{true, false} {
+		snapInfo := s.installSnap(c, developerMode, before)
+		s.cmds["apparmor_parser"].ForgetCalls()
+		snapInfo = s.updateSnap(c, snapInfo, developerMode, after)
+		profile := filepath.Join(s.backend.Directory(), "snap.samba.smbd")
+		// apparmor_parser was used to reload the profile because snap version is
+		// inside the generated policy.
+		c.Check(s.cmds["apparmor_parser"].Calls(), DeepEquals, []string{
+			fmt.Sprintf("--replace --write-cache -O no-expr-simplify --cache-loc=/var/cache/apparmor %s", profile),
+		})
+	}
+}
+
+func (s *backendSuite) TestUpdatingSnapToOneWithMoreApps(c *C) {
+	const before = `
+name: samba
```

Comment:
I did a pass that removes most of those but actually leaves this pair and a similar (but swapped) pair in ``TestUpdatingSnapTOOneWithFewerApps`` since they are unique to those tests. I'll do a second pass that makes those shared as well.